### PR TITLE
Implement flux for the categories page

### PR DESCRIPTION
**Pull request by krazemon**
I'm going to put comments on this to explain a bunch of stuff, but let's just plan on going over it together because I think that that will be the best way to explain it anyway.

### CODE REVIEW

File: js/actions/CategoriesActionCreator.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+var RepcoinAppDispatcher = require('../dispatcher/RepcoinAppDispatcher.js');
+var RepcoinConstants = require('../constants/RepcoinConstants.js');
+var ActionTypes = RepcoinConstants.ActionTypes;
+
+module.exports = {
+  getCategories: function() {
+    RepcoinAppDispatcher.handleViewAction({
+      type: ActionTypes.GET_CATEGORIES
+    });
+
```
Comment (krazemon) [inline js/actions/CategoriesActionCreator.js]: This is how we fetch data from the server in our new Flux architecture. First, we emit an event that will let everyone know that we're about to retrieve data from our server. Then, we call our api and ask it to retrieve data. This api will, once it's recieved the data, create it's own action indicating the data's been received.

File: js/api.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+var $ = require('jquery');
+
+var ServerActionCreator = require('./actions/RepcoinServerActionCreator.js');
+
+module.exports = {
+  getCategories: function() {
+    $.ajax({
+      url: '/api/categories',
+      success: ServerActionCreator.receiveCategories,
```
Comment (krazemon) [inline js/api.js]: This is where we make our server requests. Basically, we ask the server for data, and, when we get it back successfully, create an action which is going to propagate to all of the stores dependent on this data, in this case, the CategoriesStore.

File: js/components/CategoriesPage.jsx [hunk 3]
```diff
@@ -4,104 +4,47 @@ var $ = require('jquery');
...
 var CategoriesList = require('./CategoriesList.jsx');
 var CategoriesPageFilter = require('./CategoriesPageFilter.jsx');
+var CategoriesActionCreator = require('../actions/CategoriesActionCreator.js');
+var CategoriesStore = require('../stores/CategoriesStore.js');
 var Footer = require('./Footer.jsx');
 var React = require('react');
 var strings = require('../lib/strings_utils.js');
 var Toolbar = require('./Toolbar.jsx');
 
+function getStateFromStores() {
```
Comment (krazemon) [inline js/components/CategoriesPage.jsx]: This pattern of creating the function outside the React class comes from the Facebook people as does the '_' convention.

I have faith that there's a reason they use it.

File: js/dispatcher/RepcoinAppDispatcher.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
+var assign = require('object-assign');
+var Dispatcher = require('flux').Dispatcher;
+
+var RepcoinConstants = require('../constants/RepcoinConstants.js');
+
+var PayloadSources = RepcoinConstants.PayloadSources;
+
```
Comment (krazemon) [inline js/dispatcher/RepcoinAppDispatcher.js]: This comment should explain this pretty well.

File: package.json [hunk 5]
```diff
@@ -37,7 +37,7 @@
     "mongoose-validator": "^1.0.2",
     "newrelic": "~1.14.4",
     "nodemailer": "~1.3.0",
-    "nodemon": "~1.2.1",
```
Comment (krazemon) [inline package.json]: I got tired of getting that chafey nodemon message.

File: js/components/CategoriesPage.jsx (hunk 3 above)
Comment (ritmatter) [inline js/components/CategoriesPage.jsx]: What are you referring to?

File: js/stores/CategoriesStore.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  },
+
+  /* Function for registering view's change handlers */
+  addChangeListener: function(callback) {
+    this.on(CHANGE_EVENT, callback);
+  },
+
+  /* Typically gets called in 'componentWillUnmount' in views */
+  removeChangeListener: function(callback) {
+  this.removeListener(CHANGE_EVENT, callback);
```
Comment (ritmatter) [inline js/stores/CategoriesStore.js]: indent

**Comment by ritmatter** [comment]
lets talk about this in person but it looks cool

**Pull request merged**